client/src/app/editor/file-outputs/file-outputs.component.ts

```diff
@@ -57,30 +60,64 @@ export class FileOutputsComponent implements OnChanges, OnInit {
   ngOnInit() {
     const outputFileId = this.route.snapshot.queryParamMap.get('preview');
 
-    if (outputFileId) {
-      this.hasOutput
-        .pipe(
-          filter(hasOutput => hasOutput),
-          mergeMap(() => this.dataSource.connect()),
-          mergeMap(outputFiles => outputFiles),
-          filter(outputFile => outputFile.id === outputFileId),
-          filter(outputFile => isImage(outputFile.name)),
-          take(1)
+    const executionSelected$ = this.executionId$.pipe(
+      switchMap(executionId => this.labExecutionService.executionExists(executionId).pipe(catchError(() => of(false)))),
+      filter(Boolean)
+    );
+
+    const executionsAdded$ = this.labExecutionService.observeExecutionsForLab(this.lab).pipe(
+      mergeMap(executions => executions),
+      filter(labExecution => labExecution.id === this.executionId),
+      distinctUntilChanged((a, b) => a.id === b.id),
+      skip(1)
+    );
+
+    merge(executionSelected$, executionsAdded$)
+      .pipe(switchMap(() => this.labExecutionService.observeExecution(this.executionId)))
+      .subscribe(execution => {
+        this.isExecuting = execution.status === ExecutionStatus.Executing;
+      });
+
+    this.isLoading$ = merge(
+      this.executionId$.pipe(mapTo(true)),
+      this.executionId$.pipe(
+        switchMap(executionId => this.outputFilesService.hasOutputFiles(executionId)),
+        delay(4000)
+      )
+    );
+
+    this.outputFiles$ = this.executionId$.pipe(
+      switchMap(() =>
+        this.outputFilesService.observeOutputFilesFromExecution(this.executionId).pipe(
+          scan((acc: OutputFile[], val: OutputFile) => [...acc, val], []),
+          auditTime(100),
```

Comment:
Why do we need audittime here?

Reply:
`auditTime` is used here to make the stagger animation work. This is a tough one to explain. It's due to how the output files are emitted. We are using `scan` and over time _extend_ or accumulate the array of output files. This means that it could emit first one file `[{ id: 1, ... }]`, which is being animated in. Then shortly after this, could be only a few milliseconds, a second item is emitted and the array looks like this `[{ id: 1, ... }, { id: 2, ... }]`. The item with `id: 1` is already animating, and now the `mat-table` sees there's another item so it will also animate that one in. However, it's not gonna remove the first one because it kind of uses `trackBy` under the hood. If you have changes in the array so quickly that this will break the stagger effect and both items will animate at the same time. However, if we bundle the items in small "time frames" using `auditTime`, we essentially wait for the emission for 100ms and then emit the array of output files. So at least all items that arrive within that time frame will be staggered in. No matter how many items are emitted, it will probably always be within that 100ms time frame.

I hope this makes it a little bit clearer.